snow/networking/blacklist/blacklist.go

```diff
@@ -0,0 +1,231 @@
+		b.pendingQueries[key] = validatorRequests
+	}
+	validatorRequests[requestID] = struct{}{}
+
+	return true
+}
+
+// RegisterResponse removes the query from pending
+func (b *queryBlacklist) RegisterResponse(validatorID ids.ShortID, requestID uint32) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if ok := b.removeQuery(validatorID, requestID); !ok {
+		return
+	}
+
+	// Reset consecutive failures on success
+	delete(b.consecutiveFailures, validatorID.Key())
+}
+
+// QueryFailed notes a failure and blacklists [validatorID] if necessary
+func (b *queryBlacklist) QueryFailed(validatorID ids.ShortID, requestID uint32) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if ok := b.removeQuery(validatorID, requestID); !ok {
+		return
+	}
+
+	key := validatorID.Key()
+	// Add a failure and blacklist [validatorID] if it has
+	// passed the threshold
+	b.consecutiveFailures[key]++
+	if b.consecutiveFailures[key] >= b.threshold {
+		b.blacklist(validatorID)
+	}
+}
+
+func (b *queryBlacklist) blacklist(validatorID ids.ShortID) {
+	if b.blacklistSet.Contains(validatorID) {
+		return
+	}
+
+	key := validatorID.Key()
+
+	// Add to blacklist times
+	b.blacklistTimes[key] = b.clock.Time().Add(b.duration)
+	b.blacklistOrder.PushBack(validatorID)
+	b.blacklistSet.Add(validatorID)
+
+	// Note: there could be a memory leak if a large number of
+	// validators were added, sampled, and then were blacklisted
+	// and never sampled again. Due to the minimum staking amount
+	// and durations this is not a realistic concern.
+	b.cleanup()
+}
+
+// blacklisted checks if [validatorID] is currently blacklisted
+// and calls cleanup if its blacklist period has elapsed
+func (b *queryBlacklist) blacklisted(validatorID ids.ShortID) bool {
+	key := validatorID.Key()
+
+	end, ok := b.blacklistTimes[key]
+	if !ok {
+		return false
+	}
+
+	if b.clock.Time().Before(end) {
+		return true
+	}
+
+	// If a blacklisted item has expired, cleanup the blacklist
+	b.cleanup()
+	return false
+}
+
+// cleanup ensures that we have not blacklisted too much stake
+// and removes anything from the blacklist whose time has expired
+func (b *queryBlacklist) cleanup() {
+	currentWeight, err := b.vdrs.SubsetWeight(b.blacklistSet)
+	if err != nil {
+		b.reset()
+		return
+	}
+
+	maxBlacklistWeight := uint64(float64(b.vdrs.Weight()) * b.maxPortion)
+
+	// Iterate over elements of the blacklist in order of expiration
+	for e := b.blacklistOrder.Front(); e != nil; e = e.Next() {
+		validatorID := e.Value.(ids.ShortID)
+		// Remove elements with the next expiration until
+		key := validatorID.Key()
+		end := b.blacklistTimes[key]
+		// Note: this creates an edge case where blacklisting a validator
+		// with a sufficient stake may clear the blacklist
+		if b.clock.Time().Before(end) && currentWeight < maxBlacklistWeight {
+			break
+		}
+
+		removeWeight, ok := b.vdrs.GetWeight(validatorID)
+		if ok {
+			newWeight, err := safemath.Sub64(currentWeight, removeWeight)
+			if err != nil {
+				b.reset()
```

Comment:
Again, I don't think this should silently fail

Reply:
Agreed